File: sample-chat-obj-c/sample-chat/Components/Authorization/SplashScreenVC.m
//
//  SplashScreenVC.m
//  samplechat
//
//  Created by Injoit on 1/29/20.
//  Copyright © 2020 Quickblox. All rights reserved.
//

#import "SplashScreenVC.h"
#import "Profile.h"
#import "Reachability.h"
#import "Log.h"
#import "AppDelegate.h"
#import "RootParentVC.h"

NSString *const DEFAULT_PASSWORD = @"quickblox";

@interface SplashScreenVC ()
//MARK: - Properties
@property (weak, nonatomic) IBOutlet UILabel *loginInfoLabel;

@end

@implementation SplashScreenVC
//MARK: - Life Cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    
    Profile *profile = [[Profile alloc] init];
    if (!profile.isFull) {
        [(RootParentVC *)[self shared].window.rootViewController showLoginScreen];
    }
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    // Reachability
    void (^updateLoginInfo)(NetworkStatus status) = ^(NetworkStatus status) {
        
        NSString *loginInfo = (status == NetworkStatusNotReachable) ?
        NSLocalizedString(@"Please check your Internet connection", nil):
        NSLocalizedString(@"Login into chat ...", nil);
        [self updateLoginInfoText:loginInfo];
        
        Profile *profile = [[Profile alloc] init];
        if (profile.isFull && status != NetworkStatusNotReachable) {
            [self loginWithFullName:profile.fullName login:profile.login password:profile.password];
        }
    };
    
    Reachability.instance.networkStatusBlock = ^(NetworkStatus status) {
        updateLoginInfo(status);
    };
    
    updateLoginInfo(Reachability.instance.networkStatus);
}

#pragma mark - Internal Methods
/**
 *  login
 */
- (void)loginWithFullName:(NSString *)fullName login:(NSString *)login password:(NSString *)password {
    [self updateLoginInfoText:@"Login with current user ..."];
    
    __weak __typeof(self)weakSelf = self;
    [QBRequest logInWithUserLogin:login
                         password:password
                     successBlock:^(QBResponse * _Nonnull response, QBUUser * _Nonnull user) {
        
        __typeof(weakSelf)strongSelf = weakSelf;
        
        [user setPassword:password];
        [Profile synchronizeUser:user];
        [strongSelf connectToChat:user];
        
    } errorBlock:^(QBResponse * _Nonnull response) {
        __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf handleError:response.error.error];
        if (response.status == QBResponseStatusCodeUnAuthorized) {
            // Clean profile
            [Profile clearProfile];
            dispatch_async(dispatch_get_main_queue(), ^{
                [(RootParentVC *)[self shared].window.rootViewController showLoginScreen];
            });
        }
    }];
}

/**
 *  connectToChat
 */
- (void)connectToChat:(QBUUser *)user {
    
    [self updateLoginInfoText:@"Login into chat ..."];
    
    __weak __typeof(self)weakSelf = self;
    
    if ([QBChat.instance isConnected]) {
        //did Login action
        dispatch_async(dispatch_get_main_queue(), ^{
            [(RootParentVC *)[self shared].window.rootViewController showDialogsScreen];
        });
    } else {
        
        [QBChat.instance connectWithUserID:user.ID
                                  password:DEFAULT_PASSWORD
                                completion:^(NSError * _Nullable error) {
            
            __typeof(weakSelf)strongSelf = weakSelf;
            
            if (error) {
                if (error.code == QBResponseStatusCodeUnAuthorized) {
                    // Clean profile
                    [Profile clearProfile];
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [(RootParentVC *)[self shared].window.rootViewController showLoginScreen];
                    });
                    
                } else {
                    [strongSelf handleError:error];
                }
            } else {
                //did Login action
                dispatch_async(dispatch_get_main_queue(), ^{
                    [(RootParentVC *)[strongSelf shared].window.rootViewController showDialogsScreen];
                });
            }
        }];
    }
}

#pragma mark - Helpers    
- (void)updateLoginInfoText:(NSString *)text {
    if ([text isEqualToString:self.loginInfoLabel.text] == NO) {
        self.loginInfoLabel.text = text;
    }
}

- (AppDelegate*)shared {
    return (AppDelegate*) [[UIApplication sharedApplication] delegate];
}

#pragma mark - Handle errors
- (void)handleError:(NSError *)error {
    NSString *infoText = error.localizedDescription;
    if (error.code == NSURLErrorNotConnectedToInternet) {
        infoText = NSLocalizedString(@"Please check your Internet connection", nil);
    }
    [self updateLoginInfoText: infoText];
}

@end

File: sample-chat-swift/sample-chat-swift/Components/Authorization/SplashScreenVC.swift
//
//  SplashScreenVC.swift
//  sample-chat-swift
//
//  Created by Injoit on 1/27/20.
//  Copyright © 2020 quickBlox. All rights reserved.
//

import UIKit
//import UserNotifications

class SplashScreenVC: UIViewController {
    //MARK: - IBOutlets
    @IBOutlet weak var loginInfoLabel: UILabel!
    
    //MARK: - Properties
    private var infoText = "" {
        didSet {
            loginInfoLabel.text = infoText
        }
    }
    
    //MARK: - Life Cycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let profile = Profile()
        if profile.isFull == false {
            DispatchQueue.main.async {
                AppDelegate.shared.rootViewController.showLoginScreen()
            }
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        //MARK: - Reachability
        let updateLoginInfo: ((_ status: NetworkConnectionStatus) -> Void)? = { [weak self] status in
            let notConnection = status == .notConnection
            let loginInfo = notConnection ? LoginConstant.checkInternet : LoginStatusConstant.intoChat
            let profile = Profile()
            if profile.isFull == true, notConnection == false {
                self?.login(fullName: profile.fullName, login: profile.login)
            }
            self?.infoText = loginInfo
        }
        
        Reachability.instance.networkStatusBlock = { status in
            updateLoginInfo?(status)
        }
        updateLoginInfo?(Reachability.instance.networkConnectionStatus())
    }
    
    //MARK: - Internal Methods
    /**
     *  login
     */
    private func login(fullName: String, login: String, password: String = LoginConstant.defaultPassword) {
        QBRequest.logIn(withUserLogin: login,
                        password: password,
                        successBlock: { [weak self] response, user in
                            guard let self = self else {
                                return
                            }
                            
                            user.password = password
                            Profile.synchronize(user)
                            self.connectToChat(user: user)
                            
                        }, errorBlock: { [weak self] response in
                            self?.handleError(response.error?.error, domain: ErrorDomain.logIn)
                            if response.status == QBResponseStatusCode.unAuthorized {
                                // Clean profile
                                Profile.clearProfile()
                                DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + .seconds(1)) {
                                    AppDelegate.shared.rootViewController.showLoginScreen()
                                }
                            }
                            
                        })
    }
    
    /**
     *  connectToChat
     */
    private func connectToChat(user: QBUUser) {
        infoText = LoginStatusConstant.intoChat
        if QBChat.instance.isConnected == true {
            //did Login action
            DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + .seconds(1)) {
                AppDelegate.shared.rootViewController.goToDialogsScreen()
            }
        } else {
            QBChat.instance.connect(withUserID: user.id,
                                    password: LoginConstant.defaultPassword,
                                    completion: { [weak self] error in
                                        guard self != nil else { return }
                                        if let error = error {
                                            if error._code == QBResponseStatusCode.unAuthorized.rawValue {
                                                // Clean profile
                                                Profile.clearProfile()
                                                DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + .seconds(1)) { AppDelegate.shared.rootViewController.showLoginScreen()
                                                }
                                            }
                                            
                                        } else {
                                            //did Login action
                                            DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + .seconds(1)) {
                                                AppDelegate.shared.rootViewController.goToDialogsScreen()
                                            }
                                        }
                                    })
        }
    }
    
    // MARK: - Handle errors
    private func handleError(_ error: Error?, domain: ErrorDomain) {
        guard let error = error else {
            return
        }
        var infoText = error.localizedDescription
        if error._code == NSURLErrorNotConnectedToInternet {
            infoText = LoginConstant.checkInternet
        }
        self.infoText = infoText
    }
}
